runtime/common/src/purchase.rs

```diff
@@ -0,0 +1,910 @@
+				ensure!(status.validity != AccountValidity::Completed, Error::<T>::AlreadyCompleted);
+				status.validity = validity;
+				Ok(())
+			})?;
+			Self::deposit_event(RawEvent::ValidityUpdated(who, validity));
+		}
+
+		/// Update the balance of a valid account.
+		///
+		/// We check tht the account is valid for a balance transfer at this point.
+		///
+		/// Origin must match the `ValidityOrigin`.
+		#[weight = 0]
+		fn update_balance(origin,
+				who: T::AccountId,
+				free_balance: BalanceOf<T>,
+				locked_balance: BalanceOf<T>,
+		) {
+			T::ValidityOrigin::ensure_origin(origin)?;
+
+			Accounts::<T>::try_mutate(&who, |status: &mut AccountStatus<BalanceOf<T>>| -> DispatchResult {
+				let max_amount = status.validity.max_amount::<T>();
+				let total_balance = free_balance.checked_add(&locked_balance).ok_or(Error::<T>::Overflow)?;
+				// TODO: Maybe remove this check, and trust the origin to be adding the right values.
+				ensure!(total_balance <= max_amount, Error::<T>::InvalidBalance);
+				status.free_balance = free_balance;
+				status.locked_balance = locked_balance;
+				Ok(())
+			})?;
+			Self::deposit_event(RawEvent::BalanceUpdated(who, free_balance, locked_balance));
+		}
+
+		/// Pay the user and complete the purchase process.
+		///
+		/// We reverify all assumptions about the state of an account, and complete the process.
+		///
+		/// Origin must match the `PaymentOrigin`.
+		#[weight = 0]
+		fn payout(origin, who: T::AccountId) {
+			T::PaymentOrigin::ensure_origin(origin)?;
+
+			// Account should not be active.
+			ensure!(system::Module::<T>::is_dead_account(&who), Error::<T>::ExistingAccount);
+
+			Accounts::<T>::try_mutate(&who, |status: &mut AccountStatus<BalanceOf<T>>| -> DispatchResult {
+				// Account has a valid status (not Invalid, Pending, or Completed)...
+				ensure!(status.validity.is_valid(), Error::<T>::InvalidAccount);
+				// The balance we are going to transfer them matches their validity status
+				let total_balance = status.free_balance.checked_add(&status.locked_balance).ok_or(Error::<T>::Overflow)?;
+				// TODO: Possibly remove this check and trust the origin to check this before they payout.
+				ensure!(total_balance <= status.validity.max_amount::<T>(), Error::<T>::InvalidBalance);
+
+				// Transfer funds from the payment account into the purchasing user.
+				// TODO: This is totally safe? No chance of reentrancy?
+				let payment_account = PaymentAccount::<T>::get();
+				T::Currency::transfer(&payment_account, &who, total_balance, ExistenceRequirement::AllowDeath)?;
+
+				if !status.locked_balance.is_zero() {
+					// Account did not exist before this point, thus it should have no existing vesting schedule.
+					// So this function should never fail, however if it does, not much we can do about it at
+					// this point.
+					let unlock_block = system::Module::<T>::block_number().saturating_add(T::VestingTime::get());
```

Comment:
I think the the `unlock_block` should be a pallet parameter, instead of the `VestingTime`. I don't think we want to enforce all payouts to happen at the same block.

Reply:
My thought is that it is that DOTs should unlock N days after we initiate the payout. Even if payouts happen over a few days, the difference to the user here will be nominal. Versus we set a fixed block date, and something happen to delay the sale, and now dots are not locked as long as before.

Maybe something in-between is allowing the payout origin to specify the lock date when calling payout, or as an additional storage item we allow to be modified.

But I think a few blocks or even days difference between people's vesting time is not an issue.

Reply:
Yeah actually, it should be an editable block number IMO (which cannot be changed after the first payout). Even though it won't change much for users, I'd rather have the fairest solution here.